### PR TITLE
BUG: Fix spectral_slab and spectral_axis with pixels

### LINKED ISSUE #924: spectral_slab crashes when spectral axis is in the unit of pixel
I will have to try to get a minimally reproducible example later. Just posting the traceback now so I don't lose it. This is using spacetelescope/jdaviz#1040 (Collapse plugin) and glue-viz/glue-astronomy#61 on a data cube without WCS (uncalibrated data).

```
.../specutils/manipulation/extract_spectral_region.py in _edge_value_to_pixel(edge_value, spectrum, order, side)
     30         else:
---> 31             index = spectrum.wcs.world_to_pixel(edge_value)
     32 

.../gwcs/api.py in world_to_pixel(self, *world_objects)
    313         """
--> 314         result = self.invert(*world_objects, with_units=True)
    315         if not utils.isnumerical(result[0]):

.../gwcs/wcs.py in invert(self, *args, **kwargs)
    507             akwargs = {k: v for k, v in kwargs.items() if k not in _ITER_INV_KWARGS}
--> 508             result = self.backward_transform(*args, **akwargs)
    509         except (NotImplementedError, KeyError):

.../astropy/modeling/core.py in __call__(self, model_set_axis, with_bounding_box, fill_value, equivalencies, inputs_map, *inputs, **new_inputs)
    397 
--> 398             new_call = make_function_with_signature(
    399                 __call__, args, kwargs, varargs='inputs', varkwargs='new_inputs')

.../astropy/modeling/core.py in __call__(self, *inputs, **kwargs)
    376                 """Evaluate this model on the supplied inputs."""
--> 377                 return super(cls, self).__call__(*inputs, **kwargs)
    378 

.../astropy/modeling/core.py in __call__(self, *args, **kwargs)
   1068         # prepare for model evaluation (overridden in CompoundModel)
-> 1069         evaluate, inputs, broadcasted_shapes, kwargs = self._pre_evaluate(*args, **kwargs)
   1070 

.../astropy/modeling/core.py in _pre_evaluate(self, *args, **kwargs)
    924         # Broadcast inputs into common size
--> 925         inputs, broadcasted_shapes = self.prepare_inputs(*args, **kwargs)
    926 

.../astropy/modeling/core.py in prepare_inputs(self, model_set_axis, equivalencies, *inputs, **kwargs)
   2013 
-> 2014         inputs = self._validate_input_units(inputs, equivalencies, inputs_map)
   2015 

.../astropy/modeling/core.py in _validate_input_units(self, inputs, equivalencies, inputs_map)
   2081                         if input_unit is dimensionless_unscaled:
-> 2082                             raise UnitsError("{0}: Units of input '{1}', {2} ({3}),"
   2083                                              "could not be converted to "

UnitsError: SpectralTabular1D: Units of input 'x', pix (unknown),could not be converted to required dimensionless input

During handling of the above exception, another exception occurred:

...

.../jdaviz/configs/default/plugins/collapse/collapse.py in vue_collapse(self, *args, **kwargs)
    108         with warnings.catch_warnings():
    109             warnings.filterwarnings('ignore', message='No observer defined on WCS')
--> 110             spec = spectral_slab(self._selected_cube, spec_min, spec_max)
    111             # Spatial-spatial image only.
    112             collapsed_spec = spec.collapse(self.selected_func.lower(), axis=-1).T  # Quantity

.../specutils/manipulation/extract_spectral_region.py in spectral_slab(spectrum, lower, upper)
    224     region = SpectralRegion(lower, upper)
    225 
--> 226     return extract_region(spectrum, region)
    227 
    228 

.../specutils/manipulation/extract_spectral_region.py in extract_region(spectrum, region, return_single_spectrum)
    140     extracted_spectrum = []
    141     for subregion in region._subregions:
--> 142         left_index, right_index = _subregion_to_edge_pixels(subregion, spectrum)
    143 
    144         # If both indices are out of bounds then return an empty spectrum

.../specutils/manipulation/extract_spectral_region.py in _subregion_to_edge_pixels(subregion, spectrum)
     77         left_reg_in_spec_unit = left_func(subregion).to(spectral_axis.unit,
     78                                                         u.spectral())
---> 79         left_index = _edge_value_to_pixel(left_reg_in_spec_unit, spectrum, order, "left")
     80 
     81     # Right/upper side of sub region

.../specutils/manipulation/extract_spectral_region.py in _edge_value_to_pixel(edge_value, spectrum, order, side)
     39 
     40     except Exception as e:
---> 41         raise ValueError(f"Bound {edge_value}, could not be converted to pixel index"
     42                          f" using spectrum's WCS. Exception: {e}")
     43 

ValueError: Bound 0.0 pix, could not be converted to pixel index using spectrum's WCS.
Exception: SpectralTabular1D: Units of input 'x', pix (unknown),
could not be converted to required dimensionless input
```

### PR DESCRIPTION
**Pull request by pllim**
This PR:

* Fix #924 
* Fix a bug where you see the following:

```python
>>> from specutils import Spectrum1D
>>> sc = Spectrum1D(flux=np.ones((2, 2, 5)) * u.nJy)
>>> sc.spectral_axis  # Expects 0-4 pix
<SpectralAxis [ 0.,  1., nan, nan, nan] pix>
```

### CODE REVIEW

**Comment by pllim** [comment]
Remote data failure unrelated.

File: specutils/tests/test_region_extract.py [hunk 1]
```diff
@@ -34,7 +34,7 @@ def test_region_simple(simulated_spectra):
 
     sub_spectrum_flux_expected = np.array(FLUX_ARRAY)
 
-    assert quantity_allclose(sub_spectrum.flux.value, sub_spectrum_flux_expected)
+    assert_quantity_allclose(sub_spectrum.flux.value, sub_spectrum_flux_expected)
```
Comment (rosteen) [inline specutils/tests/test_region_extract.py:37]: Thought I was going crazy here for a minute and there wasn't actually a diff, then I noticed the underscore 😆

**Review by rosteen — APPROVED** [review]
Thanks!

File: specutils/spectra/spectrum1d.py [hunk 2]
```diff
@@ -265,7 +265,7 @@ def __init__(self, flux=None, spectral_axis=None, wcs=None,
         elif wcs is None:
             # If no spectral axis or wcs information is provided, initialize
             # with an empty gwcs based on the flux.
-            size = len(flux) if not flux.isscalar else 1
+            size = flux.shape[-1] if not flux.isscalar else 1
```
Comment (dhomeier) [inline specutils/spectra/spectrum1d.py:268]: Wondering if the `isscalar` check is even necessary after the `flux = u.Quantity([flux])` further up.
But then there is an alternative hook where `flux = kwargs.pop('data')` might be slipping through; not sure if that is by design?

Comment (pllim) [reply specutils/spectra/spectrum1d.py:268]: 🤷

**Pull request merged**